src/classes/PS_IntegrationServiceConfig.cls

```diff
@@ -44,9 +70,7 @@ public with sharing class PS_IntegrationServiceConfig {
     private Boolean isEnabled {
         get {
             if (isEnabled == null) {
-                //Integer configCount = Database.countQuery('SELECT Count() FROM Payment_Services_Configuration__c');
-                Integer configCount = 0;//TODO: hard-coded, disable the integration service until full implementation
-                isEnabled = configCount > 0;
+                return new Configuration().keyValueMap.keySet().containsAll(REQUIRED_CONFIG_KEYS);
```

Comment:
I believe that the Configuration wrapper class is being instantiated twice by the two methods (isEnabled and hasIntegrationPermissions). Each new  configuration instance will query the config object, resulting in an extra query. What do you think of creating a method in the outer class such as the following that both methods can use to retrieve the configuration without incurring the additional query.
```
private static Configuration psConfig {
    get {
        if (psConfig == null) {
            psConfig = new Configuration();
        }
        return psConfig;
    } set;
}
```
}

Reply:
We don't query the config object when we instantiate the Configuration object. The constructor is empty. @force2b

Reply:
True. However each individual instance of Configuration has its own versions of the instance methods. So, the following two calls will create two instances of configuration and each will query the `Payment_Services_Configuration__c` object. 
- `new Configuration().keyValueMap.keySet().containsAll(REQUIRED_CONFIG_KEYS);`
- `new Configuration().lastModifiedRecord`
My main concern is the addition of two soql queries per RD record when both methods are called from triggers and batch jobs. My general proposal is to move to a static instance of the query so that there is never more than 1 query against the `Payment_Services_Configuration__c` object even if the isEnabled and hasIntegrationPermission permissions are called multiple times in the same batch operation.

Reply:
I understand, I can make changes to achieve these goals, but please keep in mind however that if the configuration records get updated in 'real-time' these methods might not return the correct values, as the results will be 'cached'.

Reply:
Only if the records were updated in the same execution context, which probably can't happen. For example, if a batch job is running and processing 2000 Recurring Donation records with a batch size of 200, the result of the query against `Payment_Services_Configuration__c` is cached for the life of that batch child iteration - 200 records only. Any static vars are reset at the start of each individual child iteration (basically at the start of the execute() method). The same would be true for Trigger executions. Effectively, static var life is very short - microseconds to a possible 2 minute max cpu time for a batch iteration.